Added attribute name caching to BindBufferToShaderAttribute

gepbird opened the pull request:
Gl.GetAttribLocation is an expensive call. It usually gets called multiple times every frame.
I implemented caching in BindBufferToShaderAttribute, so it won't call Gl.GetAttribLocation every time.
The reason why I only do this at GlMethods.BindBufferToShaderAttribute, is because I don't want to modifiy the OpenGL calls in Gl.cs.

gepbird committed:
Added attribute name caching to BindBufferToShaderAttribute

gepbird committed:
Optimized lookups for the ShaderProgram

gepbird committed:
Optimized lookups for the attribute names

gepbird committed:
Remove attribute names cache when disposing a ShaderProgram.

gepbird commented on OpenGL/Core/GlMethods.cs:
```diff
@@ -723,7 +727,20 @@ public static void BindBuffer<T>(VBO<T> buffer)
+            {
+                shaderProgramAttributeNamesCache[program.ProgramID] = new Dictionary<string, uint>();
+            }
+            Dictionary<string, uint> attributeNamesCache = shaderProgramAttributeNamesCache[program.ProgramID];
```
I've done the changes. https://github.com/giawa/opengl4csharp/pull/39/commits/22689b7e6418116d647dbd5abfdb288741c30578

gepbird commented on OpenGL/Core/GlMethods.cs:
```diff
@@ -723,7 +727,20 @@ public static void BindBuffer<T>(VBO<T> buffer)
+                attributeNamesCache[attributeName] = (uint)Gl.GetAttribLocation(program.ProgramID, attributeName);
+            }
+
+            uint location = attributeNamesCache[attributeName];
```
I've done the changes. https://github.com/giawa/opengl4csharp/pull/39/commits/31249c0adcc605cb760432860fd8f57f1cacafe2

gepbird commented:
> 
> 
> The shader program cache should be removed when `Gl.DeleteProgram` is called as the program id can then be reused by a new shader program.

Shader program cache gets removed when `ShaderProgram.Dispose` gets called. https://github.com/giawa/opengl4csharp/pull/39/commits/a4e5929d65a12acf3ed5cc9c72c3a3664a9f7a36

gepbird commented:
I'm not sure about touching Gl.cs, since it's an auto generated code and most of the code calls OpenGL calls without any changes. 
> 
> 
> The reason why I only do this at GlMethods.BindBufferToShaderAttribute, is because I don't want to modifiy the OpenGL calls in Gl.cs.

Right now, only `Gl.BindBufferToShaderAttribute` and `ShaderProgram.Dispose` calls use the caching. So if someone uses the call `Gl.GetAttribLocation` or `Gl.DeleteShader` it won't use caching.

Is it a good idea to move the caching stuff to `Gl.GetAttribLocation` and `Gl.DeleteShader`, or leave it as it is now?

giawa commented:
Sorry, it has been a while since I have looked at this codebase, so it might take me a while to get into it.  I believe the attributes are already cached (along with uniforms, etc).  This is done as part of the `Shader` constructor, and then `ShaderProgram` calls `GetParams` to put together a list of them for the entire `ShaderProgram`.

You can then access that by using the overridden getter on `ShaderProgram` like so:

`var attribute = program["myAttributeName"];`

The `Location` field of `attribute` should contain an already cached location.  So you could probably do something like so:

```
public static void BindBufferToShaderAttribute<T>(VBO<T> buffer, ShaderProgram program, string attributeName) 
    where T : struct
{
    var cachedAttribute = program[attributeName];
    if (cachedAttribute == null) throw new ArgumentException($"{attributeName} did not exist in the shader.", nameof(attributeName));
    uint location = (uint)cachedAttribute.Location;

    Gl.EnableVertexAttribArray(location);
    Gl.BindBuffer(buffer);
    Gl.VertexAttribPointer(location, buffer.Size, buffer.PointerType, true, Marshal.SizeOf(typeof(T)), IntPtr.Zero);
}
```

My main concern is that the GLSL lexer may not extract 100% of the attribute names at this time.  I think it would be worth spending some time on that code to ensure it catches everything.  Instead of using the lexer it would be awesome to go with an approach like this:  https://stackoverflow.com/questions/440144/in-opengl-is-there-a-way-to-get-a-list-of-all-uniforms-attribs-used-by-a-shade

gepbird committed:
Changed cache location from attribNamesCache to ShaderProgram's ProgramParams

gepbird commented:
All right, I removed the `attribNamesCache` and replaced it with `ShaderProgram`'s `ProgramParam` cache. https://github.com/giawa/opengl4csharp/pull/39/commits/98a90ffe4de48746a9d6e24bcf478b51232b0224

> My main concern is that the GLSL lexer may not extract 100% of the attribute names at this time. I think it would be worth spending some time on that code to ensure it catches everything. Instead of using the lexer it would be awesome to go with an approach like this: https://stackoverflow.com/questions/440144/in-opengl-is-there-a-way-to-get-a-list-of-all-uniforms-attribs-used-by-a-shade

When the `ShaderProgram` gets constructed, `GetParams` will be called. It uses the same concept as you [linked](https://stackoverflow.com/questions/440144/in-opengl-is-there-a-way-to-get-a-list-of-all-uniforms-attribs-used-by-a-shade). The `ShaderProgram` already has the attribute names, so we don't need to get the names again, we can refer to the cache.

gepbird committed:
Removed an uneccessary using

giawa commented on OpenGL/Core/GlMethods.cs:
```diff
@@ -723,7 +723,10 @@ public static void BindBuffer<T>(VBO<T> buffer)
-            uint location = (uint)Gl.GetAttribLocation(program.ProgramID, attributeName);
+            var cachedAttribute = program[attributeName];
+            if(cachedAttribute == null)
+                throw new ArgumentException($"{attributeName} did not exist in the shader.", nameof(attributeName));
```
On second thought, I think it is best to avoid throwing an exception here.  If the cachedAttribute doesn't exist we should just fallback to the old method of calling `Gl.GetAttribLocation(program.ProgramID, attributeName);`.  Throwing the exception introduces a possible regression for some use cases/projects.

gepbird replied:
All right, if `cachedAttribute` is null, it will just use `Gl.GetAttribLocation`.
https://github.com/giawa/opengl4csharp/pull/39/commits/9ce0d030838443d2d604f70f69cf077f9eb27863

gepbird committed:
Use Gl.GetAttribLocation when cachedAttribute is null instead of throwing an exception

giawa commented:
Looks good!  Thanks for the contribution
